### PR TITLE
ci: Revamp CircleCI config

### PR DESCRIPTION
**Pull request by notnoop**
This adopts pattern used by Vault, where we split CircleCI yaml config
into multiple files that get packed and translated to 2.0.

This has two motivations: First, to ease translating config to CircleCI
2.0 so it can run on Enterprise private repository.  Second and most
importantly, it also adding Enterprise specific jobs in separate files
with reduced config file merging conflict resolution.

Also, has two small changes:
* remove `build-deps-image` as it's obsolete and never used
* Avoid running backend and ui tests for `stable-website` that slow down publishing website and cutting releases.

### CODE REVIEW

File: .circleci/README.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+## Quickstart
+
+The basic workflow is:
+
+- Edit source files in `./config/`
+- When you are done, run `make ci-config` to update `./config.yml`
+- Commit this entire `.circleci` directory, including that generated file together.
+- Run `make ci-verify` to ensure the current `./config.yml` is up to date with the source.
+
+When merging this `.circleci` directory:
```
Comment (tgross) [inline .circleci/README.md]: Can we clarify this? It sounds like there's a special process for merging PRs for this directory, which shouldn't be the case.

File: .circleci/README.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+## Updating the config source
+
+Before making changes, be sure to understand the layout
+of the `./config/` file tree, as well as circleci 2.1 syntax.
+See the [Syntax and layout] section below.
+
+To update the config, you should edit, add or remove files
+in the `./config/` directory,
+and then run `make ci-config`.
+If that's successful,
```
Comment (tgross) [inline .circleci/README.md]: The line-length formatting for this file is really strange. Is there a reason it isn't wrapped at a consistent length?

File: .circleci/README.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+you should then commit every `*.yml` file in the tree rooted in this directory.
+That is: you should commit both the source under `./config/`
+and the generated file `./config.yml` at the same time, in the same commit.
+The included git pre-commit hook will help with this.
+Do not edit the `./config.yml` file directly, as you will lose your changes
+next time `make ci-config` is run.
```
Comment (tgross) [inline .circleci/README.md]: We should probably provide a pre-commit hook that detects changes in this directory and runs `make ci-config` on them.

File: .circleci/README.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+in the root of your repository.
+All path references in this README assume we are in this `.circleci` directory.
+
+The `Makefile` in this directory generates `./config.yml` in CircleCI 2.0 syntax,
+from the tree rooted at `./config/`, which contains files in CircleCI 2.0 or 2.1 syntax.
+
+
+## Quickstart
+
+The basic workflow is:
```
Comment (tgross) [inline .circleci/README.md]: This quickstart section isn't appreciably different than the "full" instructions later on, except it uses less precise language. Let's cut it.

File: .circleci/README.md (hunk 3 above)
Comment (tgross) [reply .circleci/README.md]: Correction: it looks like there's supposed to be one but it wasn't included in the PR.

**Comment by notnoop** [comment]
Good points @tgross about the doc.  I've simply removed the doc now with a link to Vault one - I'll update it in a follow up iteration.

**Pull request merged**